modules/navigation/godot_navigation_server.cpp

```diff
@@ -808,8 +814,117 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+}
+
+/////////////////////////////////////
+#ifdef MODULE_NAVIGATION_ENABLED
+void GodotNavigationServer::_query_add_processqueue(const gd::PathQueryParameters &p_query_parameters, const ObjectID &p_result_object_id) const {
+	gd::PathQueryTask *pathquery_task = new gd::PathQueryTask();
+
+	pathquery_task->query_parameters = p_query_parameters;
+	pathquery_task->result_object_id = p_result_object_id;
+
+	pathquery_process_queue.push_back(pathquery_task);
+};
+
+gd::PathQueryResult GodotNavigationServer::_query_path(const gd::PathQueryParameters &p_parameters) const {
+	gd::PathQueryResult r_query_result;
+
+	const NavMap *map = map_owner.get_or_null(p_parameters.map);
+	ERR_FAIL_COND_V(map == nullptr, r_query_result);
+
+	// run the pathfinding
+
+	if (p_parameters.pathfinding_algorithm == gd::PathfindingAlgorithm::PATHFINDING_ALGORITHM_ASTAR) {
+		r_query_result.path = map->get_path(p_parameters.start_position, p_parameters.target_position, p_parameters.optimize, p_parameters.navigation_layers);
+	} else {
+		return r_query_result;
+	}
+
+	// add path postprocessing
+
+	// add calculated path stats
+
+	if (r_query_result.path.size() > 1 && p_parameters.calculate_path_length) {
+		_calculate_path_length(p_parameters, r_query_result);
+	}
+
+	return r_query_result;
+}
+
+void GodotNavigationServer::_calculate_path_length(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const {
+	float _sum = 0.0;
+	if (p_query_result.path.size() > 1) {
+		for (int i = 1; i < p_query_result.path.size(); i++) {
+			const Vector3 &prev_path_pos = p_query_result.path[i - 1];
+			const Vector3 &path_pos = p_query_result.path[i];
+			_sum += prev_path_pos.distance_to(path_pos);
+		}
+	}
+	p_query_result.path_length = _sum;
+}
+
+void GodotNavigationServer::_process_path_query(gd::PathQueryTask *pathquery_task) {
+	pathquery_task->processed = true;
+
+	Object *queryresult_object = ObjectDB::get_instance(pathquery_task->result_object_id);
+	if (!queryresult_object) {
+		// query result object id empty, nothing to call home
+		return;
+	}
+
+	const gd::PathQueryResult _query_result = _query_path(pathquery_task->query_parameters);
+
+	// NavigationPathQueryResult2D
+	if (Object::cast_to<NavigationPathQueryResult2D>(queryresult_object) != nullptr) {
+		NavigationPathQueryResult2D *pathqueryresult2d = Object::cast_to<NavigationPathQueryResult2D>(queryresult_object);
+
+		Vector<Vector2> path_vec2;
+		path_vec2.resize(_query_result.path.size());
+		for (int i = 0; i < _query_result.path.size(); i++) {
+			path_vec2.set(i, Vector2(_query_result.path[i].x, _query_result.path[i].z));
+		}
+		pathqueryresult2d->set_path(path_vec2);
+		pathqueryresult2d->set_path_length(_query_result.path_length);
+
+		// NavigationPathQueryResult3D
+	} else if (Object::cast_to<NavigationPathQueryResult3D>(queryresult_object) != nullptr) {
+		NavigationPathQueryResult3D *pathqueryresult3d = Object::cast_to<NavigationPathQueryResult3D>(queryresult_object);
+
```

Comment:
In both of these cases, it might be better to do the cast once to a variable, then test the variable in the if. You don't know how expensive the `Object::cast_to()` call is (it looks like it reverts to RTTI in some situations).

Reply:
Agree, will move the cast for the two potential results before the if block.